### PR TITLE
Component tests

### PR DESCRIPTION
**Pull request by amirmalka**
## **Type**
enhancement, tests

___

## **Description**
- Added comprehensive component tests for node-agent, including scenarios for alert generation, resource usage monitoring, and malicious behavior simulation.
- Implemented a simulated malicious application in Go for testing various security rules.
- Developed utility functions for Kubernetes and Prometheus to facilitate component testing.
- Updated the crypto mining domains list in the rule engine for accuracy.
- Removed unnecessary debug logs from the malware manager for cleaner code.
- Introduced new Kubernetes ConfigMap templates for configuring node-agent via Helm charts.
- Added a locust file for performing load testing on the node-agent.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>component_test.go</strong><dd><code>Add Comprehensive Component Tests for Node-Agent</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/component_test.go
<li>Added comprehensive component tests for node-agent functionalities.<br> <li> Tests include scenarios like basic alert tests, all alerts from <br>malicious app, basic load activities, memory leak tests, and process <br>kill in the middle of execution.<br> <li> Utilizes utility functions from 'testutils' for creating namespaces, <br>workloads, and handling Prometheus metrics.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/245/files#diff-6a05c8f454bbbdb1b2982f6178accede9be3c69a290c31887050dcf7605ab164">+399/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>malicious.go</strong><dd><code>Implement Malicious App Simulation for Testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/images/malicious-app/malicious.go
<li>Implemented a malicious application behavior simulation for testing.<br> <li> Includes various malicious activities like unexpected process launch, <br>file access, system calls, and network behaviors.<br> <li> Supports configurable behaviors through environment variables.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/245/files#diff-d9422b33824a6af4de452ff5b876da7b7b2326241622f8b2eb3fa283cb110852">+305/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>k8s.go</strong><dd><code>Add Kubernetes Utilities for Component Tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/testutils/k8s.go
<li>Added utility functions for Kubernetes operations in tests.<br> <li> Functions include creating test workloads, executing commands in pods, <br>and handling application profiles.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/245/files#diff-703f8ff02f59b4a7413f3f1970984a6e6dbb11547ac785b6a9fa93df6dc4310c">+257/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>prometheus.go</strong><dd><code>Add Prometheus Interaction Utilities for Tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/testutils/prometheus.go
<li>Added utility functions for interacting with Prometheus in tests.<br> <li> Functions for fetching and plotting CPU and memory usage metrics.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/245/files#diff-6745e4d709efd16add0f8a5a51b425da1903aacf1f1b25796d5e2c54139397c1">+242/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>locustfile.py</strong><dd><code>Add Locust File for Load Testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/images/locustimage/locustfile.py
<li>Added a locust file for load testing with HTTP requests to a target <br>URL.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/245/files#diff-8544299538f6f74b61b2b00d2a9a0c4f8fcb199bd682bc1abb822dbbbd858bda">+10/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>r1008_crypto_mining_domain.go</strong><dd><code>Update Crypto Mining Domains List</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/ruleengine/v1/r1008_crypto_mining_domain.go
<li>Updated the list of crypto mining domains with correct domain names <br>ending with a period.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/245/files#diff-6789c9f27a5f55cdaf52ed9dbf4b560714092348f28a38ccdfe86cc64e356259">+105/-105</a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>malware_manager.go</strong><dd><code>Clean up Debug Logs in Malware Manager</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/malwaremanager/v1/malware_manager.go
<li>Removed debug log statements from various functions in the malware <br>manager.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/245/files#diff-39e99490e457120b984db82ca98c0e5227942707d1e37245ffd2287079ad89af">+4/-13</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Configuration changes
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>configmap.yaml</strong><dd><code>Add Node-Agent ConfigMap Template</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/chart/templates/node-agent/configmap.yaml
<li>Created a new ConfigMap for node-agent with various configurations <br>based on values from Helm chart.<br>

</details>
    

  </td>
  <td><a href="https://github.com/kubescape/node-agent/pull/245/files#diff-80b2d032a5b110464daa7ee374ec401e9601a5ac7c2c037a90e31b309d90745b">+40/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: pkg/ruleengine/v1/r0001_unexpected_process_launched.go [hunk 1]
```diff
@@ -72,15 +74,20 @@ func (rule *R0001UnexpectedProcessLaunched) ProcessEvent(eventType utils.EventTy
 	if !ok {
 		return nil
 	}
+
 	execPath := getExecPathFromEvent(execEvent)
+	logger.L().Info("R0001UnexpectedProcessLaunched::ProcessEvent", helpers.String("execPath", execPath), helpers.String("namespace", execEvent.GetNamespace()), helpers.String("name", execEvent.GetPod()))
```
Comment (dwertent) [inline pkg/ruleengine/v1/r0001_unexpected_process_launched.go:79]: pls remove the logs from the rule

File: pkg/objectcache/applicationprofilecache/applicationprofilecache.go [hunk 2]
```diff
@@ -216,6 +216,7 @@ func (ap *ApplicationProfileCacheImpl) addApplicationProfile(_ context.Context,
 			return
 		}
 		ap.slugToAppProfile.Set(apName, fullAP)
+		logger.L().Info("added pod to application profile cache", helpers.String("name", apName))
```
Comment (dwertent) [inline pkg/objectcache/applicationprofilecache/applicationprofilecache.go:219]: Change to debug level

File: pkg/applicationprofilemanager/v1/applicationprofile_manager.go [hunk 3]
```diff
@@ -144,6 +144,12 @@ func (am *ApplicationProfileManager) deleteResources(watchedContainer *utils.Wat
 }
 
 func (am *ApplicationProfileManager) monitorContainer(ctx context.Context, container *containercollection.Container, watchedContainer *utils.WatchedContainerData) error {
+	logger.L().Info("ApplicationProfileManager - start monitor on container",
```
Comment (dwertent) [inline pkg/applicationprofilemanager/v1/applicationprofile_manager.go:147]: Change to debug level

File: Makefile [hunk 4]
```diff
@@ -1,7 +1,7 @@
 DOCKERFILE_PATH=./build/Dockerfile
 BINARY_NAME=node-agent
 
-IMAGE?=quay.io/dwertent/$(BINARY_NAME)
+IMAGE?=quay.io/amirm_armo/$(BINARY_NAME)
```
Comment (dwertent) [inline Makefile:4]:
```suggestion
IMAGE?=quay.io/kubescape/$(BINARY_NAME)
```

File: .github/workflows/pr-created.yaml [hunk 5]
```diff
@@ -1,20 +1,20 @@
-name: pull_request_created
-on:
-  pull_request:
-    types: [opened, reopened, synchronize, ready_for_review]
-    paths-ignore:
-      - '*.md'
-      - '*.yaml'
-      - '.github/workflows/*'
+# name: pull_request_created
```
Comment (dwertent) [inline .github/workflows/pr-created.yaml:1]: ?

**Review by dwertent — CHANGES_REQUESTED** [review]

**Comment by amirmalka** [comment]
@dwertent  The PR is not ready for review...

File: pkg/utils/utils.go [hunk 6]
```diff
@@ -524,7 +524,7 @@ func CreateProcessTree(process *apitypes.Process, shimPid uint32) (*apitypes.Pro
 	}
 
 	// build the process tree
-	treeRoot, err := buildProcessTree(proc, &procfs, shimPid, *process)
+	treeRoot, err := buildProcessTree(proc, &procfs, shimPid, nil)
```
Comment (dwertent) [inline pkg/utils/utils.go:527]: Are you sure about the nil?

File: pkg/objectcache/applicationprofilecache/applicationprofilecache.go [hunk 7]
```diff
@@ -216,6 +216,7 @@ func (ap *ApplicationProfileCacheImpl) addApplicationProfile(_ context.Context,
 			return
 		}
 		ap.slugToAppProfile.Set(apName, fullAP)
+		logger.L().Debug("added pod to application profile cache", helpers.String("name", apName))
```
Comment (dwertent) [inline pkg/objectcache/applicationprofilecache/applicationprofilecache.go:219]:
```suggestion
		logger.L().Info("added pod to application profile cache", helpers.String("name", apName))
```

File: pkg/utils/utils.go [hunk 8]
```diff
@@ -533,7 +533,12 @@ func CreateProcessTree(process *apitypes.Process, shimPid uint32) (*apitypes.Pro
 }
 
 // Recursively build the process tree.
-func buildProcessTree(proc procfs.Proc, procfs *procfs.FS, shimPid uint32, processTree apitypes.Process) (*apitypes.Process, error) {
+func buildProcessTree(proc procfs.Proc, procfs *procfs.FS, shimPid uint32, processTree *apitypes.Process) (*apitypes.Process, error) {
```
Comment (dwertent) [inline pkg/utils/utils.go:536]: We can easily write bugs in this function. Is it covered with proper units?

File: tests/chart/crds/runtime-rule-binding.crd.yaml [hunk 9]
```diff
@@ -0,0 +1,167 @@
+apiVersion: apiextensions.k8s.io/v1
```
Comment (dwertent) [inline tests/chart/crds/runtime-rule-binding.crd.yaml:1]: I'm not sure about this, but we should consider on having the helm-chart repo as a submodule, this way we can have the CRD synchronized and with a single place of truth.

**Review by dwertent — APPROVED** [review]

File: pkg/utils/utils.go (hunk 6 above)
Comment (amirmalka) [reply pkg/utils/utils.go:527]: Yes, we the recursive call starts with a an empty process tree. Before this fix we had a duplicate parent-child entry

File: pkg/objectcache/applicationprofilecache/applicationprofilecache.go (hunk 7 above)
Comment (amirmalka) [reply pkg/objectcache/applicationprofilecache/applicationprofilecache.go:219]: I think you advised to change this to debug like in the network neighbors cache...

File: tests/chart/crds/runtime-rule-binding.crd.yaml (hunk 9 above)
Comment (amirmalka) [reply tests/chart/crds/runtime-rule-binding.crd.yaml:1]: In general, I would prefer to use the existing helm chart, however, it is not possible to deploy only the storage and the node-agent...(like it was possible in the past)...

File: tests/scripts/setup-component-test-cluster.sh [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  # Load the docker image into the kind cluster.
+  kind load docker-image node-agent:latest || error_exit "Failed to load docker image into kind cluster."
+fi
+
+# Check if the cluster is minikube cluster by checking current context.
+if [ "$(kubectl config current-context)" == "minikube" ]; then
+  echo "Minikube cluster detected."
+  # Load the docker image into the minikube cluster.
+  minikube image load node-agent:latest || error_exit "Failed to load docker image into minikube cluster."
+fi
```
Comment (slashben) [inline tests/scripts/setup-component-test-cluster.sh:37]: Note, you are loading here `node-agent:latest`

File: tests/scripts/setup-component-test-cluster.sh [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+kubectl wait --for=condition=Ready pod -l app.kubernetes.io/name=prometheus -n monitoring --timeout=300s || error_exit "Prometheus did not start."
+
+STORAGE_TAG=$(./storage-tag.sh)
+NODE_AGENT_TAG=$(./node-agent-tag.sh)
+
+# Install node agent chart
+helm upgrade --install kubescape ../chart --set clusterName=`kubectl config current-context` \
+    --set nodeAgent.image.tag=$NODE_AGENT_TAG \
+    --set storage.image.tag=$NODE_AGENT_TAG \
+    -n kubescape --create-namespace --wait --timeout 5m || error_exit "Failed to install node-agent chart."
```
Comment (slashben) [inline tests/scripts/setup-component-test-cluster.sh:64]: And here you install Helm with a node-agent tag that might be different from what was loaded to minikube/kind above (`node-agent:latest`)

**Pull request merged**